File: apple/LC32LongestValidParentheses.java
class Solution {
    public int longestValidParentheses(String s) {
        if(s == null || s.length() == 0) return 0;
        Stack<Integer> stack = new Stack<>();
        int answer = 0;
        int flag = -1;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                if (flag == -1) {
                    stack.push(i);
                } else {
                    stack.push(flag);
                    flag = -1;
                }
            } else {
                if (stack.isEmpty()) {
                    flag = -1;
                } else {
                    int temp = stack.pop();
                    answer = Math.max(answer, i-temp+1);
                    flag = temp;
                }
            }
        }
        return answer;
    }
}

File: apple/LC32LongestValidParentheses.py
class Solution(object):
    def longestValidParentheses(self, s):
        if not s:
            return 0
        stack = []
        answer = 0
        flag = -1
        for index, char in enumerate(s):
            if char == '(':
                if flag == -1:
                    stack.append(index)
                else:
                    stack.append(flag)
                    flag = -1
            else:
                if not stack:
                    flag = -1
                else:
                    temp = stack.pop()
                    answer = max(answer, index-temp+1)
                    flag = temp
        return answer
        
